### PR TITLE
Use modern CMake practices to set compiler and linker flags

### PR DESCRIPTION
**Pull request by galanCA**
## Description

Changes how project flags are set from using `CMAKE_CXX_FLAGS` to using `target_compile_options()`. This is done for two reasons: Keep up with modern Cmake standards and to make it easier to set flags for different OS.

## Test

No significant tests were performed in this pull request. 

## Author Progress Checklist:

- [x] Open draft pull request
    - [x] Make title clearly understandable in a standalone change log context
    - [x] Assign yourself the issue
    - [x] Add at least one label (enhancement, bug, or maintenance)
    - [ ] Link the issue(s) addressed by this PR (under "Development" in the sidebar menu) 
- [x] Make code changes (if you haven't already)
- [ ] Self-review of code
    - [ ] My code follows the style guidelines of this project
    - [ ] I have added comments to my code, particularly in hard-to-understand areas
    - [ ] I have only committed the necessary changes for this fix or feature
    - [ ] I have made corresponding changes to the documentation
    - [ ] My changes generate no new warnings
    - [ ] I have ensured that my fix is effective or that my feature works as intended by:
        - [ ] exercising the code changes in the test framework, and/or
        - [ ] manually verifying the changes (as explained in the the pull request description above)
    - [ ] My changes pass all local tests
    - [ ] My changes successfully passes CI checks
    - [ ] Add any unit test for proof and documentation.
    - [ ] Merge in main branch and address resulting conflicts and/or test failures.
- [ ] Move pull request out of draft mode and assign reviewers
- [ ] Iterate with reviewers until all changes are approved
    - [ ] Make changes in response to reviewer comments
    - [ ] Merge in main branch and address resulting conflicts and/or test failures.
    - [ ] Re-request review in GitHub

## Reviewer Checklist:

 - [ ] Read the pull request description
 - [ ] Perform a code review on GitHub
 - [ ] Confirm all CI checks pass and there are no build warnings
 - [ ] Pull, build, and run automated tests locally
 - [ ] Perform manual tests of the fix or feature locally
 - [ ] Add any review comments, if applicable
 - [ ] Submit review in GitHub as either
     - [ ] Request changes, or
     - [ ] Approve
 - [ ] Iterate with author until all changes are approved

### CODE REVIEW

**Comment by galanCA** [comment]
After giving more thought to how to set the `/MT` and `/MTd` flags for the project, I prefer using [`add_compile_options`](https://github.com/cse-sim/cse/blob/c7b8fef892e538a2d63bcf0c70572d9083ad1c36/CMakeLists.txt#L11-L12) in the top Cmake because it is less repetitive (1 place instead of in each Cmakelist) and it is one of the first places a developer will look project general flags. 

Right now it is set in [BuildFlags](https://github.com/cse-sim/cse/blob/c7b8fef892e538a2d63bcf0c70572d9083ad1c36/cmake/BuildFlags.cmake#L29) for CSE and RCDEF, and in [vendor/Cmakelist.txt](https://github.com/cse-sim/cse/blob/c7b8fef892e538a2d63bcf0c70572d9083ad1c36/vendor/CMakeLists.txt#L103-L107) and although we talked about adding a macro to hide the repetitiveness sometimes it comes to the expense of readability.

File: CMakeLists.txt [hunk 1]
```diff
@@ -8,8 +8,14 @@ set (CMAKE_MODULE_PATH "${PROJECT_SOURCE_DIR}/cmake")
 set(CMAKE_CONFIGURATION_TYPES "Debug;Release" CACHE STRING "" FORCE)
 
 include(BuildFlags)
+# Creates a multithreaded executable file. Different method to add the /MT and /MTd flags
+#add_compile_options($<$<CXX_COMPILER_ID:MSVC>:$<$<CONFIG:RELEASE>:/MT>$<$<CONFIG:DEBUG>:/MTd>>)
 #include(cotire)
```
Comment (chipbarnaby) [inline CMakeLists.txt:13]: I think #include(cotire) removed unless someone has updated info.  From what I read, cotire is obsolete.  At cotire github from 3+ years ago --

> _The functionality provided by cotire has been superseded by features added to CMake 3.16. Support for pre-compiling and unity builds is now built into CMake. Thus, there will not be any further updates or support for this project._

File: src/cnglob.h [hunk 2]
```diff
@@ -14,6 +14,17 @@
 // CSE_DLL	build "silent" CSE DLL, screen output returned via callback
 // else CSE_CONSOLE  build console app, screen output to cmd window
 
+// #defines for different operating systems
+// _WIN32			Defined for both 32-bit and 64-bit environments 1
+// _WIN64			Defined for 64-bit environments 1
```
Comment (tanaya-bigladder) [inline src/cnglob.h:19]: Maybe add "windows" to the commends for windows macros, if only for consistency. Also there's a dangling character at the ends of the lines?

File: cmake/BuildFlags.cmake [hunk 3]
```diff
@@ -9,78 +7,84 @@ if (MSVC AND NOT ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Intel"))
...
+      /Gd         # Uses the __cdecl calling convention (x86 only).
+      /Zi         #*Generates complete debugging information.
+      /GS-        # Buffers security check.
+      /MP         # Compiles multiple source files by using multiple processes.
+    >
+    # # MSVC DEBUG Flags
+    $<$<CONFIG:Debug>:
+      /D_DEBUG    #*
+      # /MDd      #*Creates a debug multithreaded DLL using MSVCRT.lib.
+      /MTd        # Creates a debug multithreaded executable file using LIBCMTD.lib.
```
Comment (tanaya-bigladder) [inline cmake/BuildFlags.cmake:48]: I know this is pulled from the existing comments, but technically the /MT and /MTd flags don't "create" a multithreaded executable, they just use the multithreaded version of MS runtime libraries.

**Review by nealkruis — APPROVED** [review]
I think this is ready.

@chipbarnaby I ported over your "debug-defines" changes to this PR.

Items still worth considering:

- Make the warning flags global
- Move to level 4 warnings (there might be a separate PR for this?)

**Review by chipbarnaby — COMMENTED** [review]
Is NDEBUG defined on MSVC release?  It seems to still be commented out, but maybe it comes from some default?

**Pull request merged**